moped-editor/src/views/projects/signalProjectTable/SignalProjectTable.js

```diff
@@ -0,0 +1,383 @@
+        onChange={(event, value) => props.onChange(value)}
+        renderInput={params => (
+          <TextField {...params} variant="standard" label={null} />
+        )}
+      />
+    ),
+  };
+
+  return (
+    <CardContent>
+      <Grid container spacing={2}>
+        <Grid item xs={12}>
+          <Page title={"signal view"} className={classes.signalsTable}>
+            <MaterialTable
+              columns={columns}
+              components={{
+                EditRow: props => (
+                  <MTableEditRow
+                    {...props}
+                    onKeyDown={e => {
+                      if (e.keyCode === 13) {
+                        // Bypass default MaterialTable behavior of submitting the entire form when a user hits enter
+                        // See https://github.com/mbrn/material-table/pull/2008#issuecomment-662529834
+                      }
+                    }}
+                  />
+                ),
+                EditField: props =>
+                  props.columnDef.customEdit ? (
+                    cellEditComponents[props.columnDef.customEdit](props)
+                  ) : (
+                    <MTableEditField {...props} />
+                  ),
+              }}
+              data={data.moped_project}
+              title={
+                <Typography variant="h2" color="primary">
+                  Signals Table
+                </Typography>
+              }
+              options={{
+                ...(data.moped_project.length < PAGING_DEFAULT_COUNT + 1 && {
+                  paging: false,
+                }),
+                search: false,
+                rowStyle: typographyStyle,
+                actionsColumnIndex: -1,
+              }}
+              localization={{
+                header: {
+                  actions: "",
+                },
+              }}
+              editable={{
+                onRowUpdate: (newData, oldData) =>
+                  new Promise((resolve, reject) => {
+                    setTimeout(() => {
+                      projectActions.update(newData, oldData);
+
+                      setTimeout(() => refetch(), 501);
+                      resolve();
+                    }, 500);
+                  }),
+              }}
+              cellEditable={{
+                cellStyle: { minWidth: "300px" },
+                onCellEditApproved: (
+                  newValue,
+                  oldValue,
+                  rowData,
+                  columnDef
+                ) => {
+                  return new Promise((resolve, reject) => {
+                    setTimeout(() => {
+                      projectActions.cellUpdate(
+                        newValue,
+                        oldValue,
+                        rowData,
+                        columnDef
+                      );
+                      setTimeout(() => refetch(), 501);
+                      resolve();
+                    }, 500);
```

Comment:
I don't think there's anything wrong here, but I am curious about why the browser should wait half a second before doing the update. Is the internal setTimeout's duration of one millisecond later intended to serialize two events in a known order?

Reply:
In Chia's defense, I believe this is the pattern we have across the app. And I think it originally came from the example/demo in the table library. In my experience testing the app, if you reduce the timeout, you sometimes hit a race condition where the data update to the server is slower and the data in the db gets updated but the data rendered in the table is the old value. Worth more triage to understand more fully, but for now, the half second delay doesn't cause any bugs.

Reply:
A promise wrapped in a promise triggered by two trivial time variables that may not work if the internet is bad or if the servers are slow?

@chiaberry This is old code (from back in May) but you may want to look here:
https://github.com/cityofaustin/atd-moped/commit/715500cb704d83647446837184c1c7616f902fbf#diff-42256b926efe0131d06eacb259af2b962b743165a68c9607c19c0e59ae6dbc00R249-R253

And here:
https://github.com/cityofaustin/atd-moped/commit/715500cb704d83647446837184c1c7616f902fbf#diff-42256b926efe0131d06eacb259af2b962b743165a68c9607c19c0e59ae6dbc00R291-R297

In short, mutations are promises, why not just return the mutation with a refetch attached to it?

<img width="1655" alt="2021-11-11_14-41-50" src="https://user-images.githubusercontent.com/5282430/141366110-f192e999-d25c-4ba4-9c48-4df7680c8475.png">

Reply:
Y'all are right. I'm fixing it now. We already have this specific pattern with the timeouts in moped in a few places where we use Material Table, ill rewrite it to match the example you linked Sergio, thanks.